app/src/main/java/at/roteskreuz/stopcorona/model/repositories/ReportingRepository.kt

```diff
@@ -406,6 +422,60 @@ data class PersonalData(
     val tanSuccessfullyRequested: Boolean = false
 )
 
+data class Validity(
+    val rollingStartIntervalNumber: Int,
+    val rollingPeriod: Int?
+) : Comparable<Validity> {
+
+    /**
+     * Order validity by end of the interval. If [rollingPeriod] is not set assume validity for the
+     * whole day.
+     *
+     * This is not formally correct but good enough for all use cases
+     */
+    override fun compareTo(other: Validity): Int {
+        val rollingEndIntervalNumber = rollingStartIntervalNumber + (rollingPeriod ?: 144)
+        val otherRollingEndIntervalNumber =
+            with(other) { rollingStartIntervalNumber + (rollingPeriod ?: 144) }
+
+        return rollingEndIntervalNumber.compareTo(otherRollingEndIntervalNumber)
+    }
+
+    /**
+     * Special equals operator which ignores the [rollingPeriod] if it is not available on either
+     * object.
+     *
+     * This is required for old keys in the database where the rolling period was not stored
+     */
+    override fun equals(other: Any?): Boolean {
+        if (this === other) return true
+        if (other !is Validity) return false
+
+        if (rollingStartIntervalNumber != other.rollingStartIntervalNumber) return false
+        if (rollingPeriod == other.rollingPeriod) return true
+        // For old keys without a rolling period, ignore the rollingPeriod
+        if (rollingPeriod == null || other.rollingPeriod == null) return true
```

Comment:
If one of the Validity has `rollingPeriod` null do we want to consider them equal or only when both `rollingPeriod`s are null?

Reply:
The idea goes as follows:
- If we have a key uploaded with the previous version of the app it already has a password but no rollingPeriod in the DB.
- If we re-upload the same key later  (e.g. Y->R, Y->G, ...) with the new version of the app we will look up the password for that key with what ever rolling period it got assigned. This must return the already assigned key. No matter what the rolling period was.